go/billing/tasks.py

```diff
@@ -357,6 +357,36 @@ def generate_chunks(item_iter, items_per_chunk=1000, sep="\n"):
     return archive
 
 
+@task
+def gen_statement_then_archive(account_id, from_date, to_date, delete=True):
+    """Task to generate a monthly statement for the given account,
+    then archive the account.
+    """
+    generate_monthly_statement(account_id, from_date, to_date)
+    archive_transactions(account_id, from_date, to_date, delete)
+
+
+@task()
+def gen_statement_then_archive_monthly(months_ago=1):
+    """Spawn sub-tasks to generate a monthly statement
+    then archive each account.
+    """
+    from_date, to_date = month_range(months_ago=months_ago)
+
+    account_list = Account.objects.exclude(
+        statement__type=Statement.TYPE_MONTHLY,
+        statement__from_date=from_date,
+        statement__to_date=to_date)
+
+    account_list = account_list.exclude(
+        transactionarchive__from_date=from_date,
+        transactionarchive__to_date=to_date)
```

Comment:
Yup, that was my thinking too. I've written tests to cover both cases (the case where the statement exists and the case where the transaction archive exists). This does however mean that if the task broke halfway, this task would exclude it if run again, even though either the statement model or transaction archive model are erroneous or non-existant.